### PR TITLE
Implement online upgrades.

### PR DESCRIPTION
**Pull request by mikalstill**
This was pulled out of #1224 for being in reality separate and quite useful.

### CODE REVIEW

File: shakenfist/daemons/cluster.py [hunk 1]
```diff
@@ -61,7 +52,7 @@ def _cluster_wide_cleanup(self, last_loop_run):
         LOG.info('Running cluster maintenance')
 
         # Cleanup soft deleted objects
-        for objtype in OBJECT_NAMES:
+        for objtype in OBJECT_NAMES_TO_CLASSES:
             for _, objdata in etcd.get_all(objtype, None):
                 try:
                     obj = OBJECT_NAMES_TO_CLASSES[objtype].from_db(
```
Comment (mandoonandy) [inline shakenfist/daemons/cluster.py:58]: Setting `OBJECT_NAMES_TO_CLASSES` to contain actual classes causes circular references and leads to multiple code locations holding the same data.

Maybe change this to utilise `globals()`?
```
obj = globals()[objtype].from_db(
```
The classes in `cluster.py` will have be to imported directly 
```
from shakenfist.instance import Instance
```

**Review by mandoonandy — CHANGES_REQUESTED** [review]
I like the idea.

Suggested a method to avoid the circular dependency, but it may not be palatable.

File: shakenfist/metrics.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+        metrics = get_node_metrics([])
+        for possible_objname in OBJECT_NAMES:
+            minimum = None
+            for entry in metrics:
+                ver = entry.get('object_version_%s' % possible_objname)
+                if ver:
+                    if not minimum:
+                        minimum = ver
+                    else:
```
[the remark this answers is not in the document]
Comment (mikalstill) [reply shakenfist/metrics.py:66]: Yet again you find a language feature I don't know about.

[the remark this answers is not in the document]
Comment (mikalstill) [reply shakenfist/metrics.py:66]: Done.

File: shakenfist/daemons/cluster.py (hunk 1 above)
Comment (mikalstill) [reply shakenfist/daemons/cluster.py:58]: I don't want the class here though, just its name. I feel like I could do something terrible with importlib in baseobjectmapping.py, but I wasn't particularly happy with that option and bravely ran away.

File: shakenfist/metrics.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+
+def get_active_node_metrics():
+    return get_node_metrics([node_active_states_filter])
+
+
+VERSION_CACHE = None
+VERSION_CACHE_AGE = 0
+
+# This doesn't use OBJECT_NAME_TO_CLASSES because of circular imports.
```
[the remark this answers is not in the document]
Comment (mikalstill) [reply shakenfist/metrics.py:42]: Replied below.

File: shakenfist/instance.py [hunk 4]
```diff
@@ -123,6 +125,13 @@ class Instance(dbo):
     METADATA_KEY_AFFINITY = 'affinity'
 
     def __init__(self, static_values):
+        upgraded, static_values = self.upgrade(static_values)
```
[the remark this answers is not in the document]
Comment (mikalstill) [reply shakenfist/instance.py:128]: Fair. Done.

File: shakenfist/baseobject.py [hunk 5]
```diff
@@ -117,15 +118,14 @@ def _db_create(cls, object_uuid, metadata):
...
-        # NOTE(mikal): we don't do upgrades inflight. They are assumed to have
-        # been done as part of the upgrade process.
         o = etcd.get(cls.object_type, None, object_uuid)
         if not o:
             return None
 
         if o.get('version', 0) != cls.current_version:
-            raise exceptions.BadObjectVersion(
-                'Unknown version - %s: %s' % (cls.object_type, o))
+            if not cls.upgrade_supported:
```
[the remark this answers is not in the document]
Comment (mikalstill) [reply shakenfist/baseobject.py:126]: I think once an object moves to upgrades_supported, all subsequent versions should have to support online upgrades. Given the upgrades are always just adding a field, I think that will be ok? We can revisit if we ever find a case where that's not true?

**Review by mandoonandy — APPROVED** [review]
LGTM

File: shakenfist/daemons/cluster.py (hunk 1 above)
Comment (mandoonandy) [reply shakenfist/daemons/cluster.py:58]: You do need the class here, in order to call `.from_db()`

File: shakenfist/baseobject.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mandoonandy) [reply shakenfist/baseobject.py:126]: I think that's fair.

**Pull request merged**